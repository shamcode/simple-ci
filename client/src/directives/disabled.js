export default class {
    constructor() {
        this.node = null;
    }

    bind( node ) {
        this.node = node;
    }

    unbind() {
        this.node = null;
    }

    update( value ) {
        this.node.disabled = value;
    }
}
