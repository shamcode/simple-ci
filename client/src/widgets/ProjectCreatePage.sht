{% import ErrorsPanel from './ErrorsPanel.sht' %}
{% import ProjectCreateForm from './ProjectCreateForm' %}

<ErrorsPanel errors={{errors}}/>

<ProjectCreateForm
        dataSaving={{dataSaving}}
        onSubmit={{this.createProject.bind( this )}}
/>