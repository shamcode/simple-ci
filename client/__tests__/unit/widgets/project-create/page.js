import { DI } from 'sham-ui';
import ProjectCreatePage from '../../../../src/widgets/project-create/page';
import disabled from '../../../../src/directives/disabled';
import directives from '../../../../src/directives/event-listener';
import renderer from 'sham-ui-test-helpers';

const flushPromises = () => new Promise( resolve => setImmediate( resolve ) );

afterEach( () => {
    DI.bind( 'router', null );
    jest.resetModules();
    jest.clearAllMocks();
} );

it( 'renders correctly', () => {
    DI.bind( 'router', {
        generate: jest.fn().mockReturnValueOnce( '/' )
    } );

    const meta = renderer( ProjectCreatePage, {
        directives: {
            disabled,
            ...directives
        }
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'create', () => {
    DI.bind( 'router', {
        generate: jest.fn().mockReturnValueOnce( '/' ),
        navigate: jest.fn()
    } );
    const createProjectMock = jest.fn();
    DI.bind( 'store', {
        createProject: createProjectMock.mockReturnValueOnce( Promise.resolve() )
    } );

    const { widget: { container } } = renderer( ProjectCreatePage, {
        directives: {
            disabled,
            ...directives
        }
    } );
    const formData = {
        name: 'Test name',
        cwd: 'test cwd'
    };
    container.querySelector( '[name="name"]' ).value = formData.name;
    container.querySelector( '[name="cwd"]' ).value = formData.cwd;
    container.querySelector( '[type="submit"]' ).click();

    expect( createProjectMock.mock.calls ).toHaveLength( 1 );
    expect( createProjectMock.mock.calls[ 0 ] ).toHaveLength( 1 );

    const data = createProjectMock.mock.calls[ 0 ][ 0 ];
    expect( Object.keys( data ) ).toEqual( [ 'name', 'cwd' ] );
    expect( data.name ).toBe( formData.name );
    expect( data.cwd ).toBe( formData.cwd );
} );

it( 'create fail', async () => {
    expect.assertions( 3 );

    DI.bind( 'router', {
        generate: jest.fn().mockReturnValueOnce( '/' ),
        navigate: jest.fn()
    } );
    const createProjectMock = jest.fn();
    DI.bind( 'store', {
        createProject: createProjectMock.mockReturnValueOnce( Promise.reject() )
    } );

    const meta = renderer( ProjectCreatePage, {
        directives: {
            disabled,
            ...directives
        }
    } );
    const formData = {
        name: 'Test name',
        cwd: 'test cwd'
    };
    const { widget: { container } } = meta;
    container.querySelector( '[name="name"]' ).value = formData.name;
    container.querySelector( '[name="cwd"]' ).value = formData.cwd;
    container.querySelector( '[type="submit"]' ).click();

    await flushPromises();

    expect( createProjectMock.mock.calls ).toHaveLength( 1 );
    expect( createProjectMock.mock.calls[ 0 ] ).toHaveLength( 1 );
    expect( meta.toJSON() ).toMatchSnapshot();
} );
