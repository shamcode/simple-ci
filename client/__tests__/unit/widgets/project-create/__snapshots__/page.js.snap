// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create fail 1`] = `
Object {
  "ID": "widget",
  "constructorOptions": Object {
    "directives": Object {
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
    },
  },
  "html": "
<ul class=\\"breadcrumb\\">
  <li>
    <a href=\\"/\\" class=\\"\\">Projects</a>
  </li>
  <!--if-->
  <!--if-->
  <li>Create project</li>
</ul>
<!--BreadCrumb-->
<ul class=\\"errors-panel\\">
  <li>Create project fail!</li>
</ul>
<!--if-->
<!--ErrorsPanel-->
<div class=\\"project-create-form\\">
  <form>
    <div class=\\"group\\">
      <input name=\\"name\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Name</label>
    </div>
    <div class=\\"group\\">
      <input name=\\"cwd\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Current Work Directory</label>
    </div>
    <div class=\\"group\\">
      <button type=\\"submit\\"><span>Create</span></button>
    </div>
  </form>
</div>
<!--ProjectForm-->
",
  "nodes": Array [
    Object {
      "ID": undefined,
      "constructorOptions": null,
      "html": null,
      "nodes": null,
      "options": null,
    },
    Object {
      "ID": undefined,
      "constructorOptions": null,
      "html": null,
      "nodes": null,
      "options": null,
    },
    Object {
      "ID": undefined,
      "constructorOptions": null,
      "html": null,
      "nodes": null,
      "options": null,
    },
  ],
  "options": Object {
    "actionSequence": Array [
      "render",
      "bindEvents",
    ],
    "dataSaving": false,
    "directives": Object {
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
    },
    "errors": Array [
      "Create project fail!",
    ],
    "renderAsync": false,
    "renderDependence": Array [],
    "types": Array [],
  },
}
`;

exports[`renders correctly 1`] = `
Object {
  "ID": "widget",
  "constructorOptions": Object {
    "directives": Object {
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
    },
  },
  "html": "
<ul class=\\"breadcrumb\\">
  <li>
    <a href=\\"/\\" class=\\"\\">Projects</a>
  </li>
  <!--if-->
  <!--if-->
  <li>Create project</li>
</ul>
<!--BreadCrumb-->
<!--if-->
<!--ErrorsPanel-->
<div class=\\"project-create-form\\">
  <form>
    <div class=\\"group\\">
      <input name=\\"name\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Name</label>
    </div>
    <div class=\\"group\\">
      <input name=\\"cwd\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Current Work Directory</label>
    </div>
    <div class=\\"group\\">
      <button type=\\"submit\\"><span>Create</span></button>
    </div>
  </form>
</div>
<!--ProjectForm-->
",
  "nodes": Array [
    Object {
      "ID": undefined,
      "constructorOptions": null,
      "html": null,
      "nodes": null,
      "options": null,
    },
    Object {
      "ID": undefined,
      "constructorOptions": null,
      "html": null,
      "nodes": null,
      "options": null,
    },
    Object {
      "ID": undefined,
      "constructorOptions": null,
      "html": null,
      "nodes": null,
      "options": null,
    },
  ],
  "options": Object {
    "actionSequence": Array [
      "render",
      "bindEvents",
    ],
    "dataSaving": false,
    "directives": Object {
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
    },
    "errors": Array [],
    "renderAsync": false,
    "renderDependence": Array [],
    "types": Array [],
  },
}
`;
